File: src/main/java/me/alegian/thaumcraft7/impl/init/data/worldgen/tree/trunk/GreatwoodTrunkPlacer.java
package me.alegian.thaumcraft7.impl.init.data.worldgen.tree.trunk;

import com.google.common.collect.Lists;
import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import me.alegian.thaumcraft7.impl.init.registries.deferred.T7TrunkPlacerTypes;
import net.minecraft.core.BlockPos;
import net.minecraft.util.Mth;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.LevelSimulatedReader;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.levelgen.feature.configurations.TreeConfiguration;
import net.minecraft.world.level.levelgen.feature.foliageplacers.FoliagePlacer;
import net.minecraft.world.level.levelgen.feature.trunkplacers.GiantTrunkPlacer;
import net.minecraft.world.level.levelgen.feature.trunkplacers.TrunkPlacerType;

import java.util.List;
import java.util.function.BiConsumer;

public class GreatwoodTrunkPlacer extends GiantTrunkPlacer {
  public static final MapCodec<GreatwoodTrunkPlacer> CODEC = RecordCodecBuilder.mapCodec(
      builder -> trunkPlacerParts(builder).apply(builder, GreatwoodTrunkPlacer::new)
  );

  public GreatwoodTrunkPlacer(int pBaseHeight, int pHeightRandA, int pHeightRandB) {
    super(pBaseHeight, pHeightRandA, pHeightRandB);
  }

  @Override
  protected TrunkPlacerType<?> type() {
    return T7TrunkPlacerTypes.GREATWOOD_TRUNK_PLACER_TYPE.get();
  }

  @Override
  public List<FoliagePlacer.FoliageAttachment> placeTrunk(
      LevelSimulatedReader pLevel,
      BiConsumer<BlockPos, BlockState> pBlockSetter,
      RandomSource pRandom,
      int pFreeTreeHeight,
      BlockPos pPos,
      TreeConfiguration pConfig
  ) {
    List<FoliagePlacer.FoliageAttachment> list = Lists.newArrayList();
    list.addAll(super.placeTrunk(pLevel, pBlockSetter, pRandom, pFreeTreeHeight, pPos, pConfig));

    for (int i = pFreeTreeHeight - 2; i > pFreeTreeHeight / 4; i -= 1) {
      float f = pRandom.nextFloat() * (float) (Math.PI * 2);
      int j = 0;
      int k = 0;

      for (int l = 0; l < 6; l++) {
        j = (int) (1.5F + Mth.cos(f) * (float) l);
        k = (int) (1.5F + Mth.sin(f) * (float) l);
        BlockPos blockpos = pPos.offset(j, i - 3 + l / 2, k);
        this.placeLog(pLevel, pBlockSetter, pRandom, blockpos, pConfig);
        if (l == 6 - 1) list.add(new FoliagePlacer.FoliageAttachment(blockpos.above(3), 0, false));
      }
    }

    return list;
  }
}

File: src/main/java/me/alegian/thavma/impl/init/data/worldgen/tree/trunk/GreatwoodTrunkPlacer.kt
package me.alegian.thavma.impl.init.data.worldgen.tree.trunk

import com.mojang.serialization.codecs.RecordCodecBuilder
import me.alegian.thavma.impl.init.registries.deferred.T7TrunkPlacerTypes.GREATWOOD
import net.minecraft.core.BlockPos
import net.minecraft.util.Mth
import net.minecraft.util.RandomSource
import net.minecraft.world.level.LevelSimulatedReader
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.levelgen.feature.configurations.TreeConfiguration
import net.minecraft.world.level.levelgen.feature.foliageplacers.FoliagePlacer.FoliageAttachment
import net.minecraft.world.level.levelgen.feature.trunkplacers.GiantTrunkPlacer
import net.minecraft.world.level.levelgen.feature.trunkplacers.TrunkPlacerType
import java.util.function.BiConsumer

class GreatwoodTrunkPlacer(pBaseHeight: Int, pHeightRandA: Int, pHeightRandB: Int) : GiantTrunkPlacer(pBaseHeight, pHeightRandA, pHeightRandB) {
    override fun type(): TrunkPlacerType<*> {
        return GREATWOOD.get()
    }

    override fun placeTrunk(
        pLevel: LevelSimulatedReader,
        pBlockSetter: BiConsumer<BlockPos, BlockState>,
        pRandom: RandomSource,
        pFreeTreeHeight: Int,
        pPos: BlockPos,
        pConfig: TreeConfiguration
    ): List<FoliageAttachment> {
        val list = ArrayList<FoliageAttachment>()
        list.addAll(super.placeTrunk(pLevel, pBlockSetter, pRandom, pFreeTreeHeight, pPos, pConfig))

        var currY = pFreeTreeHeight - 2
        while (currY > pFreeTreeHeight / 4) {
            repeat(2) {
                branch(pRandom, pPos, currY, pLevel, pBlockSetter, pConfig, list)
            }
            currY--
        }

        return list
    }

    private fun branch(pRandom: RandomSource, pPos: BlockPos, currY: Int, pLevel: LevelSimulatedReader, pBlockSetter: BiConsumer<BlockPos, BlockState>, pConfig: TreeConfiguration, list: ArrayList<FoliageAttachment>) {
        val f = pRandom.nextFloat() * (Math.PI * 2).toFloat()
        var j: Int
        var k: Int

        val branchLength = pRandom.nextIntBetweenInclusive(4, 7)
        for (l in 0..branchLength) {
            j = (Mth.cos(f) * l).toInt()
            k = (Mth.sin(f) * l).toInt()
            val blockpos = pPos.offset(j, currY + l / 2 - 3, k)
            this.placeLog(pLevel, pBlockSetter, pRandom, blockpos, pConfig)
            if (l == branchLength) list.add(FoliageAttachment(blockpos.above(3), 0, false))
        }
    }

    companion object {
        val CODEC = RecordCodecBuilder.mapCodec { builder -> trunkPlacerParts(builder).apply(builder, ::GreatwoodTrunkPlacer) }
    }
}
